### PR TITLE
enable ENET on imx8mp_evk Cortex-A Core

### PR DESCRIPTION
**Pull request by JiafeiPan**
This patchset includes the following updates:
1. Add new PHY driver for realtek rtl8211f
2. Multiple enhancement for ENET driver eth_nxp_enet.c
    1) add RGMII support
    2) add 1G speed support
    3) Enable hal driver cache maintain optionally.
    4) configure MII speed and Duplex according to PHY configuration
    5) Enable MMIO mapping for ENET module driver, and then the other driver, such as MAC, MDIO and PTP, can reuse the same mapped address.
4. Enable ENET on imx8mp_evk Cortex-A Core platform:
   1) Add device dts nodes in SoC and Board dts, but they are disabled by default and can be enabled when it is to be used.
   2) Enable ENET clock in CCM driver
   3) Add pinctr for ENET
   4) Add rdc for ENET
   5) enable cache driver for imx8mp cortex-A Core
   6) enable direct mapping for imx8mp_evk platform.
   7) Add dts overlay in sample: zephyr/samples/net/zperf, it works well on imx8mp_evk Cortex-A platform.

### CODE REVIEW

**Comment by decsny** [comment]
so, it seems like you may have duplicated some work that @stanislav-poboril did on #69207 ...

**Comment by JiafeiPan** [comment]
> so, it seems like you may have duplicated some work that @stanislav-poboril did on #69207 ...

Very sorry to hear that, is there any method to avoid this case? 
And #69207 is duplicated with my PR with part of patch, if we merge one of PR, the other one need to rebase.

But I found #69207 is pending for a month, not sure and how we handle this situation? thanks.

**Comment by JiafeiPan** [comment]
> > so, it seems like you may have duplicated some work that @stanislav-poboril did on #69207 ...
> 
> Very sorry to hear that, is there any method to avoid this case? And #69207 is duplicated with my PR with part of patch, if we merge one of PR, the other one need to rebase.
> 
> But I found #69207 is pending for a month, not sure how we handle this situation? thanks.

As #69207 has been merged, rebased this PR, please review, thanks.

**Review by pdgendt — DISMISSED** [review]

**Review by pdgendt — DISMISSED** [review]

**Review by dleach02 — DISMISSED** [review]

**Comment by dleach02** [comment]
@JiafeiPan the HAL PR is merged. Please update your west manifest

**Comment by JiafeiPan** [comment]
> @JiafeiPan the HAL PR is merged. Please update your west manifest

Updated west.yml, thanks.

**Comment by pdgendt** [comment]
Doesn't the HAL commit need to go first?

**Comment by decsny** [comment]
put the hal change in the commit where it's needed, don't make a separate commit

**Comment by JiafeiPan** [comment]
> put the hal change in the commit where it's needed, don't make a separate commit

merged west.yml update in the commit "soc: imx8mm/n/p: enable cache driver for Cortex-A Core"

**Comment by JiafeiPan** [comment]
> Doesn't the HAL commit need to go first?

Yes, thanks.

**Comment by JiafeiPan** [comment]
And rebased.

**Comment by JiafeiPan** [comment]
The twister build failure is a general issue for all platforms to build tests/kernel/poll/, it can be fixed by https://github.com/zephyrproject-rtos/zephyr/pull/74126

File: drivers/ethernet/nxp_enet/eth_nxp_enet.c [hunk 1]
```diff
@@ -955,8 +959,8 @@ BUILD_ASSERT(NXP_ENET_PHY_MODE(DT_DRV_INST(n)) != NXP_ENET_RGMII_MODE ||		\
 				.txBdStartAddrAlign = nxp_enet_##n##_tx_buffer_desc,	\
 				.rxBufferAlign = nxp_enet_##n##_rx_buffer[0],		\
 				.txBufferAlign = nxp_enet_##n##_tx_buffer[0],		\
-				.rxMaintainEnable = true,				\
-				.txMaintainEnable = true,				\
+				.rxMaintainEnable = driver_cache_maintain,			\
+				.txMaintainEnable = driver_cache_maintain,			\
```
Comment (pdgendt) [inline drivers/ethernet/nxp_enet/eth_nxp_enet.c:963]: Nit: alignment

Comment (JiafeiPan) [reply drivers/ethernet/nxp_enet/eth_nxp_enet.c]: Fixed and rebased, thanks.

**Review by pdgendt — DISMISSED** [review]

File: boards/nxp/imx8mp_evk/imx8mp_evk-pinctrl.dtsi
Comment (decsny) [inline boards/nxp/imx8mp_evk/imx8mp_evk-pinctrl.dtsi:1]: please squash the pinctrl into the board dts commits for both of the platforms

File: dts/bindings/ethernet/nxp,enet1g.yaml [hunk 2]
```diff
@@ -13,3 +13,7 @@ properties:
 
   clocks:
     required: true
+
+  rdc:
+    type: int
+    description: Set the RDC permission for this peripheral
```
Comment (decsny) [inline dts/bindings/ethernet/nxp,enet1g.yaml:19]: please provide the meaning of the abbreviation in the description at least

File: boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53_defconfig [hunk 3]
```diff
@@ -13,6 +13,7 @@ CONFIG_ICACHE_LINE_SIZE_DETECT=y
 
 # Zephyr Kernel Configuration
 CONFIG_XIP=n
+CONFIG_KERNEL_DIRECT_MAP=y
```
Comment (decsny) [inline boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53_defconfig:16]: move before enablement where its needed or squash also

File: samples/net/zperf/boards/imx8mp_evk_mimx8ml8_a53.overlay
Comment (decsny) [inline samples/net/zperf/boards/imx8mp_evk_mimx8ml8_a53.overlay:1]: recommend to squash as well, a separate commit doesnt seem necessary

**Review by decsny — COMMENTED** [review]
sorry for late review

File: boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts [hunk 4]
```diff
@@ -34,6 +34,44 @@
...
+
+&enet_mdio {
+	status = "disabled";
+	pinctrl-0 = <&pinmux_mdio>;
+	pinctrl-names = "default";
+	phy: phy@0 {
+		compatible = "realtek,rtl8211f";
+		reg = <1>;
+		status = "disabled";
+		/* fixed-link = "1000BASE-T Full-Duplex"; */
```
[the remark this answers is not in the document]
Comment (JiafeiPan) [reply boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts:73]: Removed as current PHY driver can't support fixed link ( my original phy driver support it, so will submit this feature late).

File: drivers/ethernet/nxp_enet/eth_nxp_enet.c [hunk 5]
```diff
@@ -957,8 +960,8 @@ BUILD_ASSERT(NXP_ENET_PHY_MODE(DT_DRV_INST(n)) != NXP_ENET_RGMII_MODE ||		\
 				.txBdStartAddrAlign = nxp_enet_##n##_tx_buffer_desc,	\
 				.rxBufferAlign = nxp_enet_##n##_rx_buffer[0],		\
 				.txBufferAlign = nxp_enet_##n##_tx_buffer[0],		\
-				.rxMaintainEnable = true,				\
-				.txMaintainEnable = true,				\
+				.rxMaintainEnable = driver_cache_maintain,		\
+				.txMaintainEnable = driver_cache_maintain,		\
```
[the remark this answers is not in the document]
Comment (JiafeiPan) [reply drivers/ethernet/nxp_enet/eth_nxp_enet.c:965]: Fixed, thanks.

File: boards/nxp/imx8mp_evk/imx8mp_evk-pinctrl.dtsi
Comment (JiafeiPan) [reply boards/nxp/imx8mp_evk/imx8mp_evk-pinctrl.dtsi:1]: Done.

File: boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53_defconfig (hunk 3 above)
Comment (JiafeiPan) [reply boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53_defconfig:16]: Done

File: dts/bindings/ethernet/nxp,enet1g.yaml (hunk 2 above)
Comment (JiafeiPan) [reply dts/bindings/ethernet/nxp,enet1g.yaml:19]: Fixed, thanks.

File: samples/net/zperf/boards/imx8mp_evk_mimx8ml8_a53.overlay
Comment (JiafeiPan) [reply samples/net/zperf/boards/imx8mp_evk_mimx8ml8_a53.overlay:1]: Done.

File: drivers/ethernet/nxp_enet/eth_nxp_enet.c [hunk 6]
```diff
@@ -65,7 +65,7 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 #elif defined(CONFIG_SOC_SERIES_RW6XX)
 #define ETH_NXP_ENET_UNIQUE_ID	(OCOTP->OTP_SHADOW[46])
 #else
-#error "Unsupported SOC"
+#define ETH_NXP_ENET_UNIQUE_ID 0xFFFFFF
```
[the remark this answers is not in the document]
Comment (JiafeiPan) [reply drivers/ethernet/nxp_enet/eth_nxp_enet.c:68]: I added checking in nxp_enet_unique_mac() to block this ID will be used for MAC address.

**Comment by JiafeiPan** [comment]
Rebased.

File: boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts [hunk 7]
```diff
@@ -34,6 +34,44 @@
...
+ * &enet_mac {
+ *		status = "okay";
+ * };
+ * &enet_mdio {
+ *		status = "okay";
+ * };
+ * &phy {
+ *		status = "okay";
+ * };
+ */
```
[the remark this answers is not in the document]
Comment (JiafeiPan) [reply boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts]: Created an overlay file: boards/nxp/imx8mp_evk/dts/nxp,enet1g.overlay

**Review by dleach02 — DISMISSED** [review]

File: boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts (hunk 7 above)
[the remark this answers is not in the document]
Comment (JiafeiPan) [reply boards/nxp/imx8mp_evk/imx8mp_evk_mimx8ml8_a53.dts]: > Created an overlay file: boards/nxp/imx8mp_evk/dts/nxp,enet1g.overlay

I enabled all realted dts nodes in board dts as currently only single ethernet port is supported.

**Comment by JiafeiPan** [comment]
rebased.

**Review by pdgendt — APPROVED** [review]

**Review by dleach02 — APPROVED** [review]

**Pull request merged**